src/kpro_lib.erl

```diff
@@ -0,0 +1,301 @@
+-module(kpro_lib).
+
+-export([ copy_bytes/2
+        , data_size/1
+        , decode/2
+        , encode/2
+        , find/3
+        , get_prelude_schema/2
+        , get_req_schema/2
+        , get_rsp_schema/2
+        , get_ts_type/2
+        , keyfind/3
+        , now_ts/0
+        , ok_pipe/1
+        , ok_pipe/2
+        , parse_endpoints/2
+        , update_map/4
+        , with_timeout/2
+        ]).
+
+-include("kpro_private.hrl").
+
+-define(IS_BYTE(I), (I>=0 andalso I<256)).
+
+-type primitive_type() :: kpro:primitive_type().
+-type primitive() :: kpro:primitive().
+-type count() :: non_neg_integer().
+
+%%%_* APIs =====================================================================
+
+%% @doc Function pipeline. The fist function takes no args, all succeeding
+%% functions take one arg. All functions should retrun either
+%% `ok' | `{ok, Result}' | `{error, Reason}'. `Result' is the input arg of
+%% the next function (or the result of pipeline).
+%% NOTE: If a funcition returns `ok' the next should be an arity-0 fun.
+%% Any `{error, Reason}' return value would cause the pipeline to abort.
+%% NOTE: The pipe funcions are delegated to an agent process to evaluate
+%%       only exceptions and process links are propagated back to caller
+%%       other side-effects like monitor references are not handled.
+ok_pipe(FunList, Timeout) ->
+  with_timeout(fun() -> do_ok_pipe(FunList) end, Timeout).
+
+%% @doc Same as `ok_pipe/2' with `infinity' as default timeout.
+ok_pipe(FunList) ->
+  ok_pipe(FunList, infinity).
+
+%% @doc Parse comma separated endpoints in a string into a list of
+%% `{Host::string(), Port::integer()}' pairs.
+%% Endpoints may or may not start with protocol prefix (non case sensitive):
+%% `PLAINTEXT://', `SSL://', `SASL_PLAINTEXT://' or `SASL_SSL://'.
+%% The first arg is to filter desired endpoints from parse result.
+-spec parse_endpoints(kpro:protocol() | undefined, string()) ->
+        [kpro:endpoint()].
+parse_endpoints(Protocol, Str) ->
+  Eps0 = string:tokens(Str, ",\n"),
+  L = [parse_endpoint(string:to_lower(Ep)) || Ep <- Eps0, Ep =/= ""],
+  [EP || {P, EP} <- L, Protocol =:= P].
```

Comment:
Nit: the filtering in the 2nd list comprehension could be folded into the first, avoiding an intermediate list